base/compiler/abstractinterpretation.jl

```diff
@@ -3288,10 +3294,12 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
             # Process non control-flow statements
             (; changes, rt, exct) = abstract_eval_basic_statement(interp,
                 stmt, currstate, frame)
-            if exct !== Union{}
-                update_exc_bestguess!(interp, exct, frame)
-            end
             if !has_curr_ssaflag(frame, IR_FLAG_NOTHROW)
+                if exct !== Union{}
+                    update_exc_bestguess!(interp, exct, frame)
+                    # TODO: assert that these conditions match. For now, we assume the `nothrow` flag
+                    # to be correct, but allow the exct to be an over-approximation.
+                end
```

Comment:
In the future, I think the exct modeling should encompass the nothrow modeling. But I agree with you that this approach might be the best for now. We'd seize opportunities to enhance exct modeling as they arise.

Reply:
Yes, we should eventually delete the nothrow effect in favor of this, but I don't think there's any urgency.